File: python/76_minimum_window_substring.py
class Solution:
    def isTinS(self, s_freq: {}, t_freq: {}) -> bool:
        print(s_freq, t_freq)
        for ch in t_freq:
            if ch not in s_freq:
                return False
            if t_freq[ch] > s_freq[ch]:
                return False    
        return True
    
    def minWindow(self, s: str, t: str) -> str:
        if len(t) > len(s):
            return ""
        if len(t) == 1:
            if t in s:
                return t
            return ""
        freq_t = {}
        for ch in t:
            if ch in freq_t:
                freq_t[ch] += 1
            else:
                freq_t[ch] = 1
        freq_s = {}
        l, r = 0, 1
        freq_s[s[l]] = 1
        if s[r] in freq_s:
            freq_s[s[r]] += 1
        else:
            freq_s[s[r]] = 1
        res = s
        window_present = False
        while l <= r and r < len(s):
            print("lookinng in:: ", s[l:r+1])
            if self.isTinS(freq_s, freq_t):
                window_present = True
                temp = s[l:r+1]
                if len(temp) < len(res):
                    res = temp
                freq_s[s[l]] -= 1
                l += 1
            else:
                r += 1
                if r < len(s):
                    if s[r] in freq_s:
                        freq_s[s[r]] += 1
                    else:
                        freq_s[s[r]] = 1
        if window_present:
            return res
        return ""

sol = Solution()
inps = input()
inpt = input()
print(sol.minWindow(inps, inpt)) 

File: golang/76_minimum_window_substring.go
package main

import (
	"fmt"
	"strings"
)

func isTinS(freqS map[string]int, freqT map[string]int) bool {
	for key, tCount := range freqT {
		if sCount, ok := freqS[key]; ok {
			if tCount > sCount {
				return false
			}
		} else {
			return false
		}
	}
	return true
}

func minWindow(s string, t string) string {
	if len(t) > len(s) {
		return ""
	}
	if len(t) == 1 {
		if strings.Index(s, t) > -1 {
			return t
		}
		return ""
	}
	freqT := make(map[string]int)
	for i := 0; i < len(t); i++ {
		tempCh := string(t[i])
		if _, ok := freqT[tempCh]; ok {
			freqT[tempCh]++
		} else {
			freqT[tempCh] = 1
		}
	}
	freqS := make(map[string]int)
	l, r := 0, 1
	freqS[string(s[l])] = 1
	tempStr := string(s[r])
	if _, ok := freqS[tempStr]; ok {
		freqS[tempStr]++
	} else {
		freqS[tempStr] = 1
	}
	var res string = s
	var windowFound bool = false
	for l <= r && r < len(s) {
		if isTinS(freqS, freqT) {
			windowFound = true
			temp := s[l : r+1]
			if len(temp) < len(res) {
				res = temp
			}
			freqS[string(s[l])]--
			l++
		} else {
			r++
			if r < len(s) {
				tempStr = string(s[r])
				if _, ok := freqS[tempStr]; ok {
					freqS[tempStr]++
				} else {
					freqS[tempStr] = 1
				}
			}
		}
	}
	if windowFound == true {
		return res
	}
	return ""
}

func main() {
	var inps, inpt string
	fmt.Scanln(&inps)
	fmt.Scanln(&inpt)
	fmt.Println(inps, inpt)
	fmt.Println(minWindow(inps, inpt))
}
